api/handlers.go

```diff
@@ -165,3 +171,404 @@ func (api *RecipeAPI) AddRecipeHandler(w http.ResponseWriter, req *http.Request)
+	if err != nil {
+		log.Event(ctx, "error updating recipe to mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Marshal to output the newly added recipe
+	output, err := json.Marshal(currentRecipe)
+	if err != nil {
+		log.Event(ctx, "error returned from json marshal", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set("content-type", "application/json")
+	w.Write(output)
+}
+
+//AddCodelistHandler - add a codelist in the instance of an existing recipe
+func (api *RecipeAPI) AddCodelistHandler(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+
+	//Get Update Recipe ID
+	vars := mux.Vars(req)
+	recipeID := vars["id"]
+	instanceID := vars["instance_id"]
+	logD := log.Data{"recipe_id": recipeID, "instance_id": instanceID}
+
+	// Read body
+	b, err := ioutil.ReadAll(req.Body)
+	defer req.Body.Close()
+	if err != nil {
+		log.Event(ctx, "error in reading request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Unmarshal to the shape of Codelist struct
+	var codelist recipe.CodeList
+	err = json.Unmarshal(b, &codelist)
+	if err != nil {
+		log.Event(ctx, "error returned from json unmarshal", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Generate the HRef of codelist if not given in request body as it follows a consistent pattern
+	hrefURL := "http://localhost:22400/code-lists/"
+	if codelist.HRef == "" {
+		codelist.HRef = hrefURL + instanceID
+	}
+
+	// Validation to check if all the instance fields are entered
+	err = codelist.ValidateAddCodelist()
+	if err != nil {
+		log.Event(ctx, "bad request error as invalid codelist given in request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// Get currentRecipe from ID and retrieve specific instance for codelist to be stored
+	currentRecipe, err := api.dataStore.Backend.GetRecipe(recipeID)
+	if err != nil {
+		log.Event(ctx, "error retrieving specific recipe from mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	instanceIndex := -1
+	for i, instance := range currentRecipe.OutputInstances {
+		if instance.DatasetID == instanceID {
+			instanceIndex = i
+			break
+		}
+	}
+	if instanceIndex == -1 {
+		log.Event(ctx, "error retrieving specific instance of recipe from mongo", log.ERROR, logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Append new codelist into the current codelist within the specific instance of the recipe
+	currentRecipe.OutputInstances[instanceIndex].CodeLists = append(currentRecipe.OutputInstances[instanceIndex].CodeLists, codelist)
+
+	// Update the current recipe in mongo with the updated codelist in the specific instance
+	update := bson.M{"$set": bson.M{"output_instances." + strconv.Itoa(instanceIndex): currentRecipe}}
```

Comment:
This bson statement would more typically live in our mongo packages - in order to keep your existing UpdateRecipe function generic, consider creating an AddCodeList function in mongo, which contains this update := line, and then calls UpdateRecipe function itself. It can be a lightweight wrapper, but at least then the behaviour is all encapsulated in a single package, and clearly named with its use

Reply:
This is repeated in a few places, ive flagged them all with 'See longer comment on mongo package encapsulation'

Reply:
Done in commit "move bson statements to mongo package - mongo package encapsulation"